src/asami/projection.cljc

```diff
@@ -203,6 +203,32 @@
           (throw (ex-info (str "Projection variables not found in the selected data: " missing)
                           {:missing missing :data columns})))))))
 
+(s/defn derive-pattern-and-columns :- (s/pair [s/Any] :pattern
+                                              [s/Any] :columns)
+  "Given a pattern (possibly with usage of the :as syntax), return a
+  pair of a pattern without usage of the :as syntax and renamed
+  columns as described by the :as syntax.
+
+  Example:
+
+    (derive-pattern-and-columns '[?x :data :as ?z ?y])
+    ;; => [[?x :data ?y] [?x ?z ?y]]"
+  [pattern]
+  (loop [pattern pattern
+         new-pattern []
+         renamed-columns []]
+    (if (seq pattern)
+      (let [[x y z :as xs] (take 3 pattern)
+            new-pattern (conj new-pattern x)]
+        (if (= :as y)
+          (recur (drop 3 pattern)
+                 new-pattern
+                 (conj renamed-columns (or z x)))
```

Comment:
This is safe, but if `z` is `nil`, then perhaps an exception is appropriate?

Reply:
Since this function is intended for internal use, I was thinking we could catch problems with `:as` syntax during validation (before a query is executed) and to keep concerns separate.

Reply:
OK, I was actually expecting that. But if `z` is `nil` (which should never happen) then that means there was an error in another part of the code. Hence my suggestion of an exception

Reply:
This throws an exception now.